### ISSUE TITLE
examples - update the depends_on directive to avoid "race condition" with autoconf

### ISSUE THREAD

**Issue by Sorrow81** [bunkerity/bunkerized-nginx#129]
Hello,

First of all thank you for your work, I already use bunkerized in autoconf for a static site and a php site. I am trying in vain to make a reverse proxy to serve my JS application. To do this I do this:

```bash
docker volume create nginx_conf && docker network create mynet
```

```bash
docker run -d -p 80:8080 \
           --network mynet \
           --name bunkerized \
           -v /var/www:/www:ro \
           -v nginx_conf:/etc/nginx \
           -e SERVER_NAME= \
           -e MULTISITE=yes \
           -l bunkerized-nginx.AUTOCONF \
           bunkerity/bunkerized-nginx
```
```bash
docker run -d --name autoconf \
           -v /var/run/docker.sock:/var/run/docker.sock:ro \
           -v nginx_conf:/etc/nginx \
           bunkerity/bunkerized-nginx-autoconf
```
           
At this point I tried several commands without success:
```bash
docker run --network mynet \
           --name myapi \
           -l bunkerized-nginx.SERVER_NAME=api.local \
           -l bunkerized-nginx.USE_REVERSE_PROXY=yes \
           -l bunkerized-nginx.REVERSE_PROXY_URL=/ \
           -l bunkerized-nginx.REVERSE_PROXY_HOST=http://myapi:1337 \
           tutum/hello-world
```
```bash
docker run --network mynet \
           --name myapi \
           -l bunkerized-nginx.SERVER_NAME=api.local \
           -l bunkerized-nginx.USE_REVERSE_PROXY=yes \
           -l bunkerized-nginx.REVERSE_PROXY_URL=/ \
           -l bunkerized-nginx.REVERSE_PROXY_HOST=http://myapi:1337 \
           bunkerity/bunkerized-nginx
```
```bash
docker run --network mynet \
           --name myapi \
           -l bunkerized-nginx.SERVER_NAME=api.local \
           -l bunkerized-nginx.USE_REVERSE_PROXY=yes \
           -l bunkerized-nginx.REVERSE_PROXY_URL=/ \
           -l bunkerized-nginx.REVERSE_PROXY_HOST=http://127.0.0.1:1337 \
           tutum/hello-world
```
```bash
docker run --network mynet \
           --name myapi \
           -l bunkerized-nginx.SERVER_NAME=api.local \
           -l bunkerized-nginx.USE_REVERSE_PROXY=yes \
           -l bunkerized-nginx.REVERSE_PROXY_URL=/ \
           -l bunkerized-nginx.REVERSE_PROXY_HOST=http://localhost:1337 \
           bunkerity/bunkerized-nginx
```

I can access my application locally on the machine via the url "http://localhost:1337" but cannot access it via "http://api.local".

Thank you for your help.

**Comment by Sorrow81** [comment]
Hello,

I did some research to understand a bit better how docker works. And I ended up creating a docker-compose.yaml. However I couldn't figure out how to point to my domain without having to add the port in the url.

```yml
version: "3"

services:
  mywww:
    image: bunkerity/bunkerized-nginx
    restart: always
    ports:
      - 80:8080
    volumes:
      - /www:/www:ro
      - autoconf:/etc/nginx
    environment:
      - SERVER_NAME= # must be left blank if you don't want to setup "static" conf
      - MULTISITE=yes
      - DISABLE_DEFAULT_SERVER=yes
      - USE_CLIENT_CACHE=yes
      - USE_GZIP=yes
    labels:
      - "bunkerized-nginx.AUTOCONF"

  myautoconf:
    image: bunkerity/bunkerized-nginx-autoconf
    restart: always
    volumes:
      - /var/run/docker.sock:/var/run/docker.sock:ro
      - autoconf:/etc/nginx
    depends_on:
      - mywww

  strapi:
    image: strapi/strapi:3.5.3
    restart: always
    environment:
      DATABASE_CLIENT: postgres
      DATABASE_NAME: strapi
      DATABASE_HOST: postgres
      DATABASE_PORT: 5432
      DATABASE_USERNAME: strapi
      DATABASE_PASSWORD: strapi
    volumes:
      - ./app:/srv/app
    ports:
      - "1337:1337"
    depends_on:
      - postgres
      - myautoconf
    labels:
      - "bunkerized-nginx.SERVER_NAME=toto.local"            # replace with your domain
      - "bunkerized-nginx.USE_REVERSE_PROXY=yes"
      - "bunkerized-nginx.REVERSE_PROXY_URL=/"
      - "bunkerized-nginx.REVERSE_PROXY_HOST=http://strapi"

  postgres:
    image: postgres
    environment:
      POSTGRES_DB: strapi
      POSTGRES_USER: strapi
      POSTGRES_PASSWORD: strapi
    volumes:
      - ./data:/var/lib/postgresql/data

volumes:
  autoconf:```
  
I don't understand why the application is only accessible on port 1337?

**Comment by sjpi** [comment]
I have been struggling with this as well for a few days. No luck on resolving it. I assume it is something I am missing. 

Did you make any progress on this? We seem to have the same/similar issue, I am able to locally resolve the JS app but not via a public resolver. When I start messing with the config, I just break the upstream to where it can't reach the origin (502's).

If I open a shell on the proxy I am able to curl the REVERSE_PROXY_HOST if I append the port to the end (curl http://hostname:port)

**Comment by bunkerity** [comment]
@Sorrow81 : it looks like your strapi service is listening on port TCP/1337 inside the container. You then need to append the port inside the **REVERSE_PROXY_HOST** environment variable. Here is the edited compose file : 

```yaml
version: "3"

services:
  mywww:
    image: bunkerity/bunkerized-nginx
    restart: always
    ports:
      - 80:8080
    volumes:
      - /www:/www:ro
      - autoconf:/etc/nginx
    environment:
      - SERVER_NAME= # must be left blank if you don't want to setup "static" conf
      - MULTISITE=yes
      - DISABLE_DEFAULT_SERVER=yes
      - USE_CLIENT_CACHE=yes
      - USE_GZIP=yes
    labels:
      - "bunkerized-nginx.AUTOCONF"

  myautoconf:
    image: bunkerity/bunkerized-nginx-autoconf
    restart: always
    volumes:
      - /var/run/docker.sock:/var/run/docker.sock:ro
      - autoconf:/etc/nginx
    depends_on:
      - mywww

  strapi:
    image: strapi/strapi:3.5.3
    restart: always
    environment:
      DATABASE_CLIENT: postgres
      DATABASE_NAME: strapi
      DATABASE_HOST: postgres
      DATABASE_PORT: 5432
      DATABASE_USERNAME: strapi
      DATABASE_PASSWORD: strapi
    volumes:
      - ./app:/srv/app
    # redirecting the port is facultative because bunkerized-nginx is used as a reverse proxy
    ports:
      - "1337:1337"
    depends_on:
      - postgres
      - myautoconf
    labels:
      - "bunkerized-nginx.SERVER_NAME=toto.local"            # replace with your domain
      - "bunkerized-nginx.USE_REVERSE_PROXY=yes"
      - "bunkerized-nginx.REVERSE_PROXY_URL=/"
      - "bunkerized-nginx.REVERSE_PROXY_HOST=http://strapi:1337"

  postgres:
    image: postgres
    environment:
      POSTGRES_DB: strapi
      POSTGRES_USER: strapi
      POSTGRES_PASSWORD: strapi
    volumes:
      - ./data:/var/lib/postgresql/data

volumes:
  autoconf:
```

**Comment by bunkerity** [comment]
Hi @Sorrow81,

I have another hypothesis related to a "race condition" where autoconf don't get the event about the strapi service. Can you try the following compose :

```yaml
version: "3"

services:
  mywww:
    image: bunkerity/bunkerized-nginx
    restart: always
    ports:
      - 80:8080
    volumes:
      - /www:/www:ro
      - autoconf:/etc/nginx
    environment:
      - SERVER_NAME= # must be left blank if you don't want to setup "static" conf
      - MULTISITE=yes
      - DISABLE_DEFAULT_SERVER=yes
      - USE_CLIENT_CACHE=yes
      - USE_GZIP=yes
    labels:
      - "bunkerized-nginx.AUTOCONF"

  myautoconf:
    image: bunkerity/bunkerized-nginx-autoconf
    restart: always
    volumes:
      - /var/run/docker.sock:/var/run/docker.sock:ro
      - autoconf:/etc/nginx
    depends_on:
      - mywww
      - strapi

  strapi:
    image: strapi/strapi:3.5.3
    restart: always
    environment:
      DATABASE_CLIENT: postgres
      DATABASE_NAME: strapi
      DATABASE_HOST: postgres
      DATABASE_PORT: 5432
      DATABASE_USERNAME: strapi
      DATABASE_PASSWORD: strapi
    volumes:
      - ./app:/srv/app
    # redirecting the port is facultative because bunkerized-nginx is used as a reverse proxy
    #ports:
      #- "1337:1337"
    depends_on:
      - postgres
    labels:
      - "bunkerized-nginx.SERVER_NAME=toto.local"            # replace with your domain
      - "bunkerized-nginx.USE_REVERSE_PROXY=yes"
      - "bunkerized-nginx.REVERSE_PROXY_URL=/"
      - "bunkerized-nginx.REVERSE_PROXY_HOST=http://strapi:1337"

  postgres:
    image: postgres
    environment:
      POSTGRES_DB: strapi
      POSTGRES_USER: strapi
      POSTGRES_PASSWORD: strapi
    volumes:
      - ./data:/var/lib/postgresql/data

volumes:
  autoconf:
```

Just in case, remove everything and pull the latest images before starting the containers :

```shell
docker-compose down -v
docker-compose pull
docker-compose up
```

**Comment by bunkerity** [comment]
Glad it worked for you.

Let's keep this issue open until we update the autoconf examples to add the depends_on directive.

**Comment by Sorrow81** [comment]
I understand, in the past in the autoconf it misses the backslash on the label REVERSE_PROXY_HOST

![Screenshot_20210510-222920](https://user-images.githubusercontent.com/43729314/117721197-d8ea7700-b1df-11eb-8024-9d249ce0edd0.jpg)

**Comment by ptr1337** [comment]
Just a question, if im providing several docker-compose file for different services, but dont want to inckude them into one. 
Just a shared network is enogh i think or?

**Comment by bunkerity** [comment]
@Sorrow81 : changes are now on the master branch. Thanks for your report.

@ptr1337 : 

I think that sharing the same network should do the trick. Here is a quick example.

bunkerized.yml : 
```yaml
version: "3"

services:

  mywww:
    image: bunkerity/bunkerized-nginx
    restart: always
    ports:
      - 80:8080
    volumes:
      - /www:/www:ro
      - autoconf:/etc/nginx
    environment:
      - SERVER_NAME= # must be left blank if you don't want to setup "static" conf
      - MULTISITE=yes
    labels:
      - "bunkerized-nginx.AUTOCONF"
    networks:
      - mynet

  myautoconf:
    image: bunkerity/bunkerized-nginx-autoconf
    restart: always
    volumes:
      - /var/run/docker.sock:/var/run/docker.sock:ro
      - autoconf:/etc/nginx
    depends_on:
      - mywww

volumes:
  autoconf:

networks:
  mynet:
    name: mynetname
```

service1.yml : 
```yaml
version: "3"

services:

  service1:
    image: tutum/hello-world
    networks:
      mynet:
        aliases:
          - service1
    labels:
      - "bunkerized-nginx.SERVER_NAME=service1.domain"
      - "bunkerized-nginx.USE_REVERSE_PROXY=yes"
      - "bunkerized-nginx.REVERSE_PROXY_URL=/"
      - "bunkerized-nginx.REVERSE_PROXY_HOST=http://service1"

networks:
  mynet:
    external:
      name: mynetname
```

Running : 
```shell
$ docker-compose -f bunkerized.yml up -d
... wait a little while everything is up ...
$ docker-compose -f service1.yml up -d
... wait a little until config is generated ...
$ curl -H "Host: service1.domain" -H "User-Agent: legit" http://localhost
<html>
<head>
        <title>Hello world!</title>
...
```

@sjpi : don't hesitate to open a new issue if needed.

**Comment by ptr1337** [comment]
Wow, nice changes in the last weeks! 

Now i can finally completly use it completly! 
Now only DNS Challenge missing for me, but thats not a big problem.

Thanks for your great work!
